LSM: Fix SuperBlock.Manifest.tables overflow

sentientwaffle opened the pull request:
Presently manifest compaction can fall behind, eventually overflowing `SuperBlock.Manifest.tables`.

To fix this, only store "active" table extents, instead of both active and inactive table extents.

## Context

`SuperBlock.Manifest.tables` is a hash map tracking each table's latest "extent" – that is, tracking the most recent manifest block which referenced that table.

`SuperBlock.Manifest.tables` presently tracks both:
1. "active" tables: i.e. tables that are referenced by a `ManifestLevel`
2. "inactive" tables: tables that were removed from their LSM tree, but which have not yet been garbage collected by manifest log compaction.

Right now `SuperBlock.Manifest.tables` capacity is defined as:

```zig
try tables.ensureTotalCapacity(allocator, forest_table_count_max);
```

This bound is the maximum number of active tables – it doesn't take into account any inactive tables.
And if `tables` overflows, the replica will crash.

(This is currently reproducible with the forest fuzzer because (in an unrelated bug) manifest compaction can't keep up and falls behind.)

## Fix

Two possible fixes:
1. Count inactive tables in `tables`' worst-case capacity. (~Too expensive; see below.~ Implemented in this PR.)
2. Exclude inactive tables from `tables`, but leave the capacity as-is. (~Implemented in this PR.~)

### Count inactive tables

(This approach is implemented in this PR.)

I originally thought that this approach was too expensive to be viable:

<details>
<summary>Original (incorrect!) worst-case calculation</summary>

The worst-case capacity for `tables` counting both active and inactive tables
(calculation performed with 64KiB blocks for convenience, but it doesn't make a difference in the final result):

      65.5e3 manifest blocks
    × 576    events/block
    × 8+8+8  bytes/extent
    × 2      (hash map load factor)
    = 1.7 GiB

...Ouch.

Counting only active tables:

      2.3e6  tables
    × 8+8+8  bytes/extent
    × 2      (hash map load factor)
    = 105.3 MiB

That 1.7GiB could be tuned down by increasing the manifest block load factor (i.e. compacting the manifest more frequently, reducing the maximum number of manifest blocks) but that can only go so far.

</details>

@maklad found a tighter bound! :fire:

For a particular table in the manifest, the sequence of events is:

    insert(0|1), update(0+), remove(0|1)

We only need to track the addresses of removed tables _until the corresponding "insert" is encountered_.
That, combined with the fact that the maximum number of active tables at any given moment is a constant (`tree.table_count_max`), means that there are at most `table_count_max` "unpaired" removes to count.

Those removes are tracked in a separate hashmap from the table extents – this saves a little memory, and lends itself to farther RAM optimization in the future.

### Exclude inactive tables

(This approach is ~implemented~ not implemented in this PR.)

#### Reverse-chronological manifest recovery

Why does `tables` need inactive tables right now?

The manifest recovery algorithm (`ManifestLog.open()`) is:

    tables = {}
    for manifest_block in reverse(superblock.manifest.blocks):
        for entry in reverse(manifest_block.entries):
            if tables[entry.table]:
                # This table has already been restored.
            else:
                tables[entry.table] = true
                switch entry.event:
                    insert: manifest_levels.insert(entry.table)
                    remove: {}

Manifest entries are replayed in reverse-chronological order.
If a table `T` is deleted at time `Y`, then if we encounter an insert for table `T` at time `X` (`X < Y`), we must ignore it – we must not restore it to the manifest levels.

By replaying events in reverse chronological order, recovery only needs to update the manifest levels once per surviving table.

#### Chronological manifest recovery

The downside of reverse-chronological recovery is that the replica pays (RAM) tracking those (many) inactive tables at all times, not just during recovery.

By replaying events chronologically, the inactive tables are no longer necessary:

    for manifest_block in superblock.manifest.blocks:
        for entry in manifest_block.entries:
            switch entry.event:
                insert: manifest_levels.insert(entry.table)
                remove: manifest_levels.remove(entry.table)

The tradeoff is that we must now update the manifest levels for each event in the manifest log – even events which are later superseded.

## Also in this PR

- `ManifestChecker`: Verify that the `ManifestLevel`s tables are constructed consistently across replicas and after recovering from a restart.
- `Forest.verify_tables_recovered()`: An intrusive (`constants.verify` and `TestStorage`-gated) check which reimplements the chronological manifest recovery algorithm to check `ManifestLevel`s and `SuperBlock.Manifest.tables`.
- `Forest.verify_table_extents()`: An inexpensive sanity-check between `ManifestLevel`s and `SuperBlock.Manifest.tables`.

sentientwaffle committed:
Test: Add ManifestChecker

sentientwaffle committed:
LSM: Add misc assertions

sentientwaffle committed:
LSM/ManifestLevel: Guard against UB after remove_table

Don't allow passing a `table` to `ManifestLevel.remove_table(table)` that is a pointer into the `ManifestLevel`'s own `SegmentedArray`, since the pointer will be invalidated by the remove operations. (And then it could accidentally be used by a caller higher up the stack.)

sentientwaffle committed:
LSM/Manifest: Fix SuperBlock.Manifest.tables overflow

Presently manifest compaction can fall behind, eventually overflowing `SuperBlock.Manifest.tables`.

To fix this, only store "active" table extents, instead of both active and inactive table extents.

# Context

`SuperBlock.Manifest.tables` is a hash map tracking each table's latest "extent" – that is, tracking the most recent manifest block which referenced that table.

`SuperBlock.Manifest.tables` presently tracks both:
1. "active" tables: i.e. tables that are referenced by a `ManifestLevel`
2. "inactive" tables: tables that were removed from their LSM tree, but which have not yet been garbage collected by manifest log compaction.

Right now `SuperBlock.Manifest.tables` capacity is defined as:

```zig
try tables.ensureTotalCapacity(allocator, forest_table_count_max);
```

This bound is the maximum number of active tables – it doesn't take into account any inactive tables.
And if `tables` overflows, the replica will crash.

(This is currently reproducible with the forest fuzzer because (in an unrelated bug) manifest compaction can't keep up and falls behind.)

# Fix

Two possible fixes:
1. Count inactive tables in `tables`' worst-case capacity. (Too expensive; see below.)
2. Exclude inactive tables from `tables`, but leave the capacity as-is. (Implemented in this PR.)

## Count inactive tables

The worst-case capacity for `tables` counting both active and inactive tables
(calculation performed with 64KiB blocks for convenience, but it doesn't make a difference in the final result):

      65.5e3 manifest blocks
    × 576    events/block
    × 8+8+8  bytes/extent
    × 2      (hash map load factor)
    = 1.7 GiB

...Ouch.

Counting only active tables:

      2.3e6  tables
    × 8+8+8  bytes/extent
    × 2      (hash map load factor)
    = 105.3 MiB

That 1.7GiB could be tuned down by increasing the manifest block load factor (i.e. doing more manifest compaction) but that can only go so far.

## Exclude inactive tables

(This approach is implemented in this PR.)

### Reverse-chronological manifest recovery

Why does `tables` use inactive tables right now?

The manifest recovery algorithm (`ManifestLog.open()`) is:

    tables = {}
    for manifest_block in reverse(superblock.manifest.blocks):
        for entry in reverse(manifest_block.entries):
            if tables[entry.table]:
                # This table has already been restored.
            else:
                tables[entry.table] = true
                switch entry.event:
                    insert: manifest_levels.insert(entry.table)
                    remove: {}

Manifest entries are replayed in reverse-chronological order.
If a table `T` is deleted at time `Y`, then if we encounter an insert for table `T` at time `X` (`X < Y`), we must ignore it – we must not restore it to the manifest levels.

By replaying events in reverse chronological order, recovery only needs to update the manifest levels once per surviving table.

### Chronological manifest recovery

The downside of reverse-chronological recovery is that the replica pays (RAM) tracking those (many) inactive tables at all times, not just during recovery.

By replaying events chronologically, the inactive tables are no longer necessary:

    for manifest_block in superblock.manifest.blocks:
        for entry in manifest_block.entries:
            switch entry.event:
                insert: manifest_levels.insert(entry.table)
                remove: manifest_levels.remove(entry.table)

The tradeoff is that we must now update the manifest levels for each event in the manifest log – even events which are later superseded.

sentientwaffle committed:
LSM: Verify manifest recovery and table extents

Add two new `Forest`-level verifications:

1. `verify_tables_recovered()`: An intrusive (`constants.verify` and `TestStorage`-gated) check which reimplements the original reverse-chronological manifest recovery algorithm to check `ManifestLevel`s and `SuperBlock.Manifest.tables`.
2. `verify_table_extents()`: An inexpensive sanity-check between `ManifestLevel`s and `SuperBlock.Manifest.tables`.

matklad commented:
What's the CPU perf impact here? Would be interesting to check if our original reverse chronological recovery actually helped

sentientwaffle commented:
> What's the CPU perf impact here?

After running the benchmark there are only two manifest blocks (and the second is nearly empty) so I don't have much of a sample size to pull from here :wink: 

That said, for the single full block the measurement is surprisingly stable:
- reverse-chronological is ~250μs
- chronological is ~1200μs

So chronological is about 5x slower. Though that is extremely dependent on the "load factor" of the manifest blocks -- the more often we do manifest compaction, the more chronological's time will approach reverse-chronological.

sentientwaffle committed:
Stdx: typo

sentientwaffle committed:
LSM/Manifest: Add "update" events to manifest blocks

Previously "insert" events were used for:
1. actual new tables,
2. snapshot_max updates
3. level updates (move-table)

Now "update" events take care of the latter two.

sentientwaffle committed:
LSM/Manifest: Back to reverse-chronological manifest reconstruction

For a particular table in the manifest, the sequence of events is:

    insert(0|1), update(0+), remove(0|1)

We only need to track the addresses of removed tables _until the corresponding "insert" is encountered_.
That, combined with the fact that the maximum number of active tables at any given moment is a constant (`tree.table_count_max`), means that there are at most `table_count_max` "unpaired" removes to count.

Those removes are tracked in a separate hashmap from the table extents – this saves a little memory, and enables farther RAM optimization in the future.

sentientwaffle committed:
LSM/Manifest: Move SuperBlock.Manifest.tables into ManifestLog

I think that `tables` was owned by `SuperBlock.Manifest` because there used to be many `ManifestLog`s (one per tree), which all shared the same `tables` map.
With the unified manifest, that is no longer the case.

matklad commented on src/lsm/forest.zig:
```diff
@@ -439,5 +451,145 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+                const table_removed = tables_latest.remove(forest_table_item.table.checksum);
+                assert(table_removed);
+            }
+        }
```
assert that `tables_latest` is empty after the `while`.

matklad commented on src/lsm/forest.zig:
```diff
@@ -439,5 +451,145 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+            const tree_info = tree_infos[tree_id - tree_id_range.min];
+            assert(tree_info.tree_id == tree_id);
+
+            var groove = &@field(forest.grooves, tree_info.groove_name);
```
```suggestion
            const groove = &@field(forest.grooves, tree_info.groove_name);
```

matklad replied:
Sometimes I with that Zig just didn't have const/var for locals --- with shallow const, I am skeptical whether it actually prevents bugs.

matklad commented on src/lsm/forest.zig:
```diff
@@ -439,5 +451,145 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+            assert(table_extent_counts == forest.manifest_log.tables.count());
+
+            // Verify the tables in `tables` are exactly the tables recovered by the Forest.
+            var forest_tables_iterator = ForestTableIterator{};
```
:+1: great check here!

matklad commented on src/lsm/manifest_level.zig:
```diff
@@ -261,36 +262,51 @@ pub fn ManifestLevelType(
+                    .key_max = table.key_max,
+                });
+                return .visible;
+            }
```
Hm, the API changes here looks suspicious. `level.key_range` should track key range for the latest snapshot, but here the decision to update the key_range also depends on the snapshots passed in, which feels wrong. 

In particular, if a table is not visible to the latest snapshot, but is visible to some older snapshot, we won't update the key range here, which would be wrong. 

And then there's an issue about needing to copy table-info, and the whole "caller pass parameter, we `if` on the parameter, and then the caller asserts the `if` branch". 

Can we simplify this?

It _seems_ that we have two code paths here:

* removing a "tip" table visible to the snapshot_latest
* removing a table which is not visible to snapshot_latest

Whether we have some _other_ snapshots which could see the table I think shouldn't be relevant for this function. The caller should assert that whatever is being removed isn't visible to some old, but still active, snapshot, but it should exists outside of level. 

Relatedly, signature of `remove_invisible_tables` doesn't make sense to me. It should take a slice of snapshots, rather just a single one. At the call site, we pass in `snapshot_latest` to `remove_invisible_tables`, so that, as a result, we call `Table.invisible(&.{snapshot_latest})`. But that's weird, as `Table.invisible` special-cases `snapshot_latest` already

sentientwaffle replied:
Very good point.

`table_count_visible` is in a similar boat right now -- I added a TODO to move it up from `ManifestLevel` to `Manifest`, since in that case we actually _do_ care about visibility relative to all persistent snapshots, not just `snapshot_latest`.

matklad commented on src/lsm/manifest_log.zig:
```diff
@@ -115,6 +124,25 @@ pub fn ManifestLogType(comptime Storage: type) type {
 
+        /// A map from table address to the manifest block and entry that is the latest extent
+        /// version. Used to determine whether a table should be dropped in a compaction.
+        tables: Tables,
```
Was tripped more than once by the naming here! This aren't tables actually, just extents. `extents: Extents` or `table_extents` would make more sense to me.

matklad commented on src/lsm/manifest_log.zig:
```diff
@@ -255,14 +309,29 @@ pub fn ManifestLogType(comptime Storage: type) type {
+                    assert(manifest_log.tables_removed.get(table.address) == null);
+
+                    manifest.queue_for_compaction(block_address);
+                    manifest_log.tables_removed.putAssumeCapacity(table.address, {});
```
Maybe `fetchPut` / `getOrPut` to avoid the second hashmap lookup for the sake of the assert?

matklad commented on src/lsm/manifest_log.zig:
```diff
@@ -255,14 +309,29 @@ pub fn ManifestLogType(comptime Storage: type) type {
+                    } else {
+                        var extent = manifest_log.tables.getOrPutAssumeCapacity(table.address);
+                        if (extent.found_existing) {
+                            manifest.queue_for_compaction(block_address);
```
`assert(label.event == .update)`

sentientwaffle replied:
I don't think this holds -- the event that created the existing extent was an `.update`, but the one encountered now precedes it chronologically, so it may be an `insert`.

sentientwaffle committed:
LSM/Manifest: Code review

matklad commented on src/lsm/manifest_level.zig:
```diff
@@ -160,10 +160,12 @@ pub fn ManifestLevelType(
 
         /// The range of keys in this level covered by tables visible to snapshot_latest.
-        key_range: LevelKeyRange = .{ .key_range = null },
+        key_range_latest: LevelKeyRange = .{ .key_range = null },
```
:+1: :+1: :+1:

matklad commented on src/lsm/manifest_level.zig:
```diff
@@ -256,29 +258,22 @@ pub fn ManifestLevelType(
-        /// - If the exact table does not exist in the ManifestLevel: return null.
-        /// - If the exact table exists in the ManifestLevel: remove it, and return whether the
-        ///   table is visible to any of the given snapshots or `snapshot_latest`.
+        /// Returns whether the exact table was present, (and therefore removed).
```
Maybe just assert that the table is present? I don't think removing non-existant table would make sense, as there are just no tables outside the manifest (and we key removal by the table itself, not by a key)
